django/db/migrations/executor.py

```diff
@@ -71,46 +72,91 @@ def migrate(self, targets, plan=None, fake=False, fake_initial=False):
         """
         if plan is None:
             plan = self.migration_plan(targets)
-        migrations_to_run = {m[0] for m in plan}
         # Create the forwards plan Django would follow on an empty database
         full_plan = self.migration_plan(self.loader.graph.leaf_nodes(), clean_start=True)
-        # Holds all states right before a migration is applied
-        # if the migration is being run.
+
+        all_forwards = all(not backwards for mig, backwards in plan)
+        all_backwards = all(backwards for mig, backwards in plan)
+
+        if all_forwards and not all_backwards:
+            self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
+        elif not all_forwards and all_backwards:
+            self._migrate_all_backwards(plan, full_plan, fake=fake)
+        elif plan:
```

Comment:
That code is marginally longer, but I think it's clearer. I'll leave it up to what you think is best though.